### ISSUE TITLE
Sending spoilers

### ISSUE THREAD

**Issue by Sorunome** [vector-im/riot-web#10882]
**Is your suggestion related to a problem? Please describe.**
Recently rendering spoilers has been merged into riot develop, the problem is now that there is no way of sending them!

**Describe the solution you'd like**
Sending spoilers as extended markdown like `||spoiler||` and `||reason|spoiler||` seems reasonable.

**Describe alternatives you've considered**
None, this could be a discussion!

**Additional context**
`||spoiler||` neatly aligns with existing messaging services (e.g. Discord). As for the extended `||reason|spoiler||` it seems similar enough to just a plain spoiler to implement.

**Comment by Sorunome** [comment]
A different markdown parser that is more easily customizable would greatly simplify this step, and would also allow things like `~~strikethrough~~`. Is there a chance of changing the markdown parser?

**Comment by t3chguy** [comment]
>  Is there a chance of changing the markdown parser?

I believe there is scope for this as the experience with CommonMark is kind of unsuitable IMO, especially for markdown in links which is not easily fixable from the consumer level. GFM would be a good shout but I don't see it being extensible to adding spoilers

**Comment by aaronraimist** [comment]
Links to the MSC https://github.com/Sorunome/matrix-doc/blob/soru/spoilers/proposals/2010-spoilers.md

and PR for viewing spoilers https://github.com/matrix-org/matrix-react-sdk/pull/3018

**Comment by grinapo** [comment]
* Stackexchange uses `>! spoiler` for lines - but it won't suit spoiler words.
* Discord use `||spoiler||` for inline spoilers. - that's pretty bad for anything containing multiple pipes
* Reddit uses ` [] (#s "Spoiler Filled Text")` - convoluted...
* Github uses `<details> spoiler </details>`, others `<spoiler>spoiler text</spoiler>`. This doesn't mess up normal text.

I'd go for <spoiler>xx</spoiler> or the S.E. syntax for lines, but that's just me.

**Comment by Sorunome** [comment]
custom tags are pretty much ruled out IMO as the msc specifically added an attribute to the the span tag to not need to do custom html tags

**Comment by dkasak** [comment]
The pipe syntax is really nice, IMO, and it has the additional benefit of familiarity for Discord users. I don't see multiple pipes in ordinary text as being common enough to warrant dismissing this syntax because of that. The same could be said for a lot of other Markdown syntax.

For situations where you really need pipes, there's `/plain ...` in Riot. For cases where it occurs in code, code blocks should be used instead anyway.

**Comment by SaturnSoftware** [comment]
Apparently the mobile version of riot uses /spoiler? i'm not sure why that's not implemented here already.

**Comment by grinapo** [comment]
@SaturnSoftware yes but it's only for the **whole line**, cannot spoilermark words.

**Comment by Huy-Ngo** [comment]
I've seen bots being able to do this (they are bots bridging the chatroom with discord), so I suppose there must be already a way.

**Comment by t3chguy** [comment]
Bots don't use Element.
Element can already *render* spoilers, just has no way for users to send them.

**Comment by Sorunome** [comment]
> I've seen bots being able to do this (they are bots bridging the chatroom with discord), so I suppose there must be already a way.

yes, `/html <span data-mx-spoiler="reason">spoiler content</span>`

**Comment by shmerl** [comment]
Using `/html ...` works, but if you edit that message in element, it stops rendering the spoiler. Is there way to edit it too?

**Comment by shmerl** [comment]
Is there any way to send an image as a spoiler, not just text?

**Comment by theotheroracle** [comment]
as a workaround,  you can nest an <img> tag in a <span mx-data-spoiler> tag, but you'd have to upload the image first and get the mxc, there's also no current solution for encrypted rooms .

**Comment by theotheroracle** [comment]
> Using `/html ...` works, but if you edit that message in element, it stops rendering the spoiler. Is there a way to edit it too?

you'd have to use devtools

**Comment by shmerl** [comment]
> you'd have to upload the image first and get the mxc, there's also no current solution for encrypted rooms .

How do you get an mxc for an image uploaded to somewhere on Matrix and how should it be referred to in `<img>` tag? Or is it the regular URL of the image that you can find once it's uploaded?

**Comment by shmerl** [comment]
I tested something like this with external image, but that didn't work:

```
/html <span data-mx-spoiler="spoiler"><img src='https://someurl'/></span>
```

**Comment by theotheroracle** [comment]
no, you can't use http to reference images, only mxc urls are accepted, if you click the three dots and view the source of an image, you can see the mcx:// url

**Comment by shmerl** [comment]
It worked. Blurring on images is pretty weak though. Should I open a separate bug about it?

**Comment by theotheroracle** [comment]
if there isn't one already

**Comment by shmerl** [comment]
I didn't find anything, so I opened #16412.

**Comment by robintown** [comment]
For the time being, would it be helpful to just add a `/spoiler` command like the one on Android? I could do that easily.

**Comment by shmerl** [comment]
That would be much nicer than using tags explicitly. But for images there might be a need to have an option to add that during upload.

**Comment by theotheroracle** [comment]
if there was an option to spoiler images, it would also be helpful to add an image description field

**Comment by gabrc52** [comment]
The ||spoiler|| syntax would certainly be helpful for inline spoilers.

For backwards compatibility, the client could translate it to the HTML format before sending, and a similar way to enable editing.

**Comment by t3chguy** [comment]
> For backwards compatibility, the client could translate it to the HTML format before sending

that's what it would **have** to do - clients don't render incoming markdown, given that dialects vary so much. The messages on the wire are a Matrix defined subset of HTML.

**Comment by TheSmallTeaBoi** [comment]
Any news on this?

**Comment by SimonBrandner** [comment]
You can now send a spoiler message using `/spoiler <message>`

**Comment by TheSmallTeaBoi** [comment]
what about single words?

**Comment by t3chguy** [comment]
That's why this issue is not closed yet. Any updates on that will be right here. But given it is marked as `Help Wanted` - anyone can pick it up.

**Comment by theotheroracle** [comment]
it's blocked by the switch from commonmark i think ?

**Comment by TomMettam** [comment]
Please add support for spoiler images, too.

**Comment by shmerl** [comment]
I think images should already work, but blurring can be improved. See related: #16412

**Comment by TomMettam** [comment]
@shmerl Ah right, but is there a way to do this when attaching an image? Or just linking to one?

**Comment by shmerl** [comment]
yeah, there should be some UI / syntax options for that that are currently absent.

**Comment by robintown** [comment]
Spoiler images actually need a spec change to be properly supported, since sending images inline isn't an adequate solution for encrypted rooms.

**Comment by robintown** [comment]
I've created an MSC for media spoilers, and split their implementation out into a couple of new issues: https://github.com/vector-im/element-web/issues/18061 https://github.com/vector-im/element-web/issues/18062. If you have ideas for the designs, I encourage you to drop them in there, as it'll help me have a sane first draft to present for design review once I implement this.

**Comment by Sorunome** [comment]
For images, it could very nicely together with blurhashes

**Comment by kittykat** [comment]
@Sorunome given that we have `/spoiler` now, does that satisfy your original request or do you feel that there are additional improvements to be made? (Everyone else, please feel welcome to speak up :+1: )

**Comment by SimonBrandner** [comment]
> @Sorunome given that we have `/spoiler` now, does that satisfy your original request or do you feel that there are additional improvements to be made? (Everyone else, please feel welcome to speak up +1 )

We can't do `Hello world, this is a ||spolier||` or make images (and other things) spoilers

**Comment by ShadowJonathan** [comment]
> We can't do Hello world, this is a ||spolier|| or make images (and other things) spoilers

Could you elaborate? Is this because of limitations with commonmark libraries?

**Comment by SimonBrandner** [comment]
> > We can't do Hello world, this is a ||spolier|| or make images (and other things) spoilers
> 
> Could you elaborate? Is this because of limitations with commonmark libraries?

Ah, I didn't mean that it is technically impossible (I have no idea, but I imagine it _is_ possible). I was simply saying you can't send a message like that in Element atm, afaik

**Comment by sumnerevans** [comment]
Another possible syntax is the Reddit-style:

```
This is >!spoiler text!<
```
